File: src/main/java/mt/edu/um/cf2/jgribx/grib2/ParameterCategory.java
package mt.edu.um.cf2.jgribx.grib2;

import mt.edu.um.cf2.jgribx.Logger;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class ParameterCategory
{
    private String name;
    private int value;

    public ParameterCategory(int categoryId)
    {
        name = getName(categoryId);
        value = categoryId;
    }

    protected abstract String getName(int value);

    public static List<ParameterCategory> getCategories(int discipline)
    {
        switch (discipline)
        {
            case 0:
                return Meteorological.getCategories();
            case 1:
                return Hydrological.getCategories();
            case 2:
                return LandSurface.getCategories();
            case 3:
                return SatelliteRemoteSensing.getCategories();
            case 10:
                return Oceanographic.getCategories();
            default:
                return null;
        }
    }

    public int getValue()
    {
        return value;
    }

    @Override
    public boolean equals(Object obj)
    {
        if (obj == null || !ParameterCategory.class.isAssignableFrom(obj.getClass()))
        {
            return false;
        }
        return value == ((ParameterCategory) obj).value;
    }

    @Override
    public String toString()
    {
        return name;
    }

    public static class Meteorological extends ParameterCategory
    {
        private static final Map<Integer, String> entries = new HashMap();

        static {
            entries.put(0, "TEMPERATURE");
            entries.put(1, "MOISTURE");
            entries.put(2, "MOMENTUM");
            entries.put(3, "MASS");
            entries.put(6, "CLOUD");
            entries.put(7, "THERMODYNAMIC");
            entries.put(14, "TRACE_GASES");
        }

        public Meteorological(int categoryId)
        {
            super(categoryId);
        }

        protected String getName(int value)
        {
            return entries.get(value);
        }

        public static List<ParameterCategory> getCategories()
        {
            List<ParameterCategory> list = new ArrayList<>();
            for (Integer key : entries.keySet())
            {
                Logger.println("Adding Meteorological category: " + key, Logger.DEBUG);
                list.add(new ParameterCategory.Meteorological(key));
            }
            return list;
        }
    }

    public static class Hydrological extends ParameterCategory
    {
        private static final Map<Integer, String> entries = new HashMap<>();

        static {
            entries.put(0, "BASIC");
            entries.put(1, "PROBABILITIES");
        }

        public Hydrological(int categoryId)
        {
            super(categoryId);
        }

        protected String getName(int value)
        {
            return entries.get(value);
        }

        public static List<ParameterCategory> getCategories()
        {
            List<ParameterCategory> list = new ArrayList<>();
            for (Integer key : entries.keySet())
            {
                list.add(new ParameterCategory.Hydrological(key));
            }
            return list;
        }
    }

    public static class LandSurface extends ParameterCategory
    {
        private static final Map<Integer, String> entries = new HashMap<>();

        static {
            entries.put(0, "VEGETATION_BIOMASS");
            entries.put(1, "AGRICULTURAL_AQUACULTURAL");
            entries.put(3, "SOIL");
            entries.put(4, "FIRE_WEATHER");
        }

        public LandSurface(int categoryId)
        {
            super(categoryId);
        }

        protected String getName(int value)
        {
            return entries.get(value);
        }

        public static List<ParameterCategory> getCategories()
        {
            List<ParameterCategory> list = new ArrayList<>();
            for (Integer key : entries.keySet())
            {
                list.add(new ParameterCategory.LandSurface(key));
            }
            return list;
        }
    }

    public static class SatelliteRemoteSensing extends ParameterCategory
    {
        private static final Map<Integer, String> entries = new HashMap<>();

        static {
            entries.put(0, "IMAGE_FORMAT");
            entries.put(1, "QUANTITATIVE");
        }

        public SatelliteRemoteSensing(int categoryId)
        {
            super(categoryId);
        }

        protected String getName(int value)
        {
            return entries.get(value);
        }

        public static List<ParameterCategory> getCategories()
        {
            List<ParameterCategory> list = new ArrayList<>();
            for (Integer key : entries.keySet())
            {
                list.add(new ParameterCategory.SatelliteRemoteSensing(key));
            }
            return list;
        }
    }

    public static class Oceanographic extends ParameterCategory
    {
        private static final Map<Integer, String> entries = new HashMap<>();

        static {
            entries.put(0, "WAVES");
            entries.put(1, "CURRENTS");
        }

        public Oceanographic(int categoryId)
        {
            super(categoryId);
        }

        protected String getName(int value)
        {
            return entries.get(value);
        }

        public static List<ParameterCategory> getCategories()
        {
            List<ParameterCategory> list = new ArrayList<>();
            for (Integer key : entries.keySet())
            {
                list.add(new ParameterCategory.Oceanographic(key));
            }
            return list;
        }
    }
}

File: src/main/java/mt/edu/um/cf2/jgribx/grib2/ParameterCategory.kt
package mt.edu.um.cf2.jgribx.grib2

import mt.edu.um.cf2.jgribx.Logger

/**
 * [Parameter category by product discipline](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_table4-1.shtml)
 */
abstract class ParameterCategory private constructor(internal val value: Int, private val name: String) {
	class Meteorological private constructor(categoryId: Int, name: String) : ParameterCategory(categoryId, name) {
		companion object {
			val CATEGORIES: List<ParameterCategory> = mapOf(
					0 to "TEMPERATURE",
					1 to "MOISTURE",
					2 to "MOMENTUM",
					3 to "MASS",
					6 to "CLOUD",
					7 to "THERMODYNAMIC_STABILITY_INDICES",
					14 to "TRACE_GASES")
					.entries
					.onEach { (k, _) -> Logger.debug("Adding Meteorological category: ${k}") }
					.map { (k, v) -> Meteorological(k, v) }
		}
	}

	class Hydrological private constructor(categoryId: Int, name: String) : ParameterCategory(categoryId, name) {
		companion object {
			val CATEGORIES: List<ParameterCategory> = mapOf(
					0 to "BASIC",
					1 to "PROBABILITIES")
					.entries
					.map { (k, v) -> Hydrological(k, v) }
		}
	}

	class LandSurface private constructor(categoryId: Int, name: String) : ParameterCategory(categoryId, name) {
		companion object {
			val CATEGORIES: List<ParameterCategory> = mapOf(
					0 to "VEGETATION_BIOMASS",
					1 to "AGRICULTURAL_AQUACULTURAL",
					2 to "TRANSPORTATION",
					3 to "SOIL",
					4 to "FIRE_WEATHER",
					5 to "LAND_SURFACE")
					.entries
					.map { (k, v) -> LandSurface(k, v) }
		}
	}

	class SatelliteRemoteSensing private constructor(categoryId: Int, name: String) :
			ParameterCategory(categoryId, name) {
		companion object {
			protected val ENTRIES = mapOf(
					0 to "IMAGE_FORMAT",
					1 to "QUANTITATIVE")

			val CATEGORIES: List<ParameterCategory> = ENTRIES.entries.map { (k, v) -> SatelliteRemoteSensing(k, v) }
		}
	}

	class Oceanographic private constructor(categoryId: Int, name: String) : ParameterCategory(categoryId, name) {
		companion object {
			val CATEGORIES: List<ParameterCategory> = mapOf(
					0 to "WAVES",
					1 to "CURRENTS",
					2 to "ICE")
					.entries
					.map { (k, v) -> Oceanographic(k, v) }
		}
	}

	companion object {
		fun getCategories(discipline: Int): List<ParameterCategory>? = when (discipline) {
			0 -> Meteorological.CATEGORIES
			1 -> Hydrological.CATEGORIES
			2 -> LandSurface.CATEGORIES
			3 -> SatelliteRemoteSensing.CATEGORIES
			10 -> Oceanographic.CATEGORIES
			else -> null
		}
	}

	override fun equals(other: Any?): Boolean = this === other
			|| other is ParameterCategory
			&& value == other.value
			&& name == other.name

	override fun hashCode(): Int = 31 * value + name.hashCode()

	override fun toString(): String = name
}
